File: combinatorics/resource/ST/project_5/bai_toan_4/simple_graph/eal_to_amap.cpp
#include <iostream>
#include <vector>
#include <map>
#include <sstream>

using namespace std;

// Ham chuyen danh sach ke mo rong sang ban do ke
map<int, map<int, int>> ealToAmap(const vector<vector<pair<int, int>>>& eal, int n) {
    map<int, map<int, int>> amap;

    for (int i = 0; i < n; ++i) {
        for (const auto& [j, w] : eal[i]) {
            amap[i][j] = w;  // Gan trong so vao map
            amap[j][i] = w;
        }
    }

    return amap;
}

int main() {
    int n;
    cout << "Nhap so dinh: ";
    cin >> n;
    cin.ignore();

    vector<vector<pair<int, int>>> eal(n);
    cout << "Nhap danh sach ke mo rong:\n";

    for (int i = 0; i < n; ++i) {
        cout << "Dinh " << i << " (dinh,trongso): ";
        string line;
        getline(cin, line);
        stringstream ss(line);
        string token;
        while (ss >> token) {
            int u, w;
            sscanf(token.c_str(), "(%d,%d)", &u, &w);
            eal[i].emplace_back(u, w);
        }
    }

    auto amap = ealToAmap(eal, n);

    cout << "Ban do ke (AMap):\n";
    for (const auto& [u, neighbors] : amap) {
        cout << u << ": ";
        for (const auto& [v, w] : neighbors) {
            cout << "(" << v << "," << w << ") ";
        }
        cout << endl;
    }

    return 0;
}

File: combinatorics/resource/ST/project_5/bai_toan_4/simple_graph/eal_to_amap.py
# Ham chuyen danh sach ke mo rong sang ban do ke
def eal_to_amap(eal, n):
    amap = {}  # Tao tu dien rong

    for i in range(n):
        amap[i] = {}  # Khoi tao tu dien con cho moi dinh
        for (j, w) in eal[i]:  # Duyet cac canh ke
            amap[i][j] = w  # Gan trong so vao dinh ke

    return amap

# Nhap du lieu tu ban phim
n = int(input("Nhap so dinh: "))
eal = []

print("Nhap danh sach ke mo rong:")
for i in range(n):
    line = input(f"Dinh {i} (dinh,trongso cach nhau boi space): ")
    pairs = []
    if line.strip() != "":
        tokens = line.split()
        for token in tokens:
            u, w = map(int, token.strip("()").split(","))
            pairs.append((u, w))
    eal.append(pairs)

# Chuyen doi va in ban do ke
amap = eal_to_amap(eal, n)

print("Ban do ke (AMap):")
for u in amap:
    print(f"{u}: {amap[u]}")